samples: net: mqtt: Add a sample to connect to AWS IoT core

lucasdietrich opened the pull request:
I would like to propose the addition of a new sample application to Zephyr RTOS, which implements an MQTT client that can publish messages to AWS IoT Core using the MQTT protocol. The sample showcases the implementation of key features such as establishing a TLS 1.2 connection with AWS IoT Core, subscribing to a topic, and publishing data. Additionally, this sample is designed to pass the test suite for the AWS Device Qualification Program (DQP).

This sample can be used as a starting point for developers who are interested in building their own IoT solutions using Zephyr RTOS and AWS IoT Core.

The sample has been tested on the ST NUCLEO-F429ZI board and QEMU x86. I have taken other net/cloud samples as examples, so there are similarities in parts of the code.

Please let me know if you have any questions, feedbacks or concerns.

Fixes #22866

rlubos commented on samples/net/cloud/aws_iot_mqtt/Kconfig:
```diff
@@ -0,0 +1,74 @@
+	default AWS_TEST_SUITE_NONE
+	help
+		Select the AWS Device Advisor test suite to run.
+config AWS_TEST_SUITE_NONE
+	bool "No test suite running"
+config AWS_TEST_SUITE_DQP
+	bool "Device Qualification Program (DQP)"
+	help
+		Make sure your region supports AWS Device Advisor for DQP
+config AWS_TEST_SUITE_RECV_QOS1
+	bool "Test suite for receiving QoS 1 messages"
+	help
+		For single test case "MQTT Client Puack QoS1"
```
I'd still add an empty line in between choice options, currently, it's a bit hard to read (at least that was my first impression).

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/dhcp.h:
```diff
@@ -0,0 +1,14 @@
+
+void app_dhcpv4_startup(void);
+
+#endif /* not __DHCP_H__ */
```
Why `not`?

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/dhcp.h:
```diff
@@ -0,0 +1,14 @@
+/* DHCPv4 client startup. */
+
+/*
+ * Copyright (c) 2018 Linaro Ltd
```
The copyright seems wrong.

rlubos replied:
Sounds fine.

lucasdietrich replied:
I made the changes and updated the license in `dhcp.h` and `dhcp.c`

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h:
```diff
@@ -0,0 +1,13 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define MBEDTLS_AES_ROM_TABLES
```
There's `CONFIG_MBEDTLS_AES_ROM_TABLES` for that.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h:
```diff
@@ -0,0 +1,13 @@
+
+#define MBEDTLS_HAVE_TIME
+#define MBEDTLS_HAVE_TIME_DATE
+#define MBEDTLS_PLATFORM_TIME_ALT
```
Is this really needed? The sample doesn't seem to be using `mbedtls_platform_set_time()`

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h:
```diff
@@ -0,0 +1,13 @@
+#define MBEDTLS_HAVE_TIME_DATE
+#define MBEDTLS_PLATFORM_TIME_ALT
+
+#undef MBEDTLS_SSL_KEEP_PEER_CERTIFICATE
```
Why is this needed? The generic mbedTLS config file from Zephyr does not set this config.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/dhcp.c:
```diff
@@ -0,0 +1,66 @@
+/* DHCPv4 client startup. */
+
+/*
+ * Copyright (c) 2018 Linaro Ltd
```
The copyright seems wrong.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+#include <mbedtls/memory_buffer_alloc.h>
+#endif
+
+#include <sys/socket.h>
```
It seems to me you should rather include `#include <zephyr/net/socket.h>` given that you use `zsock_*` APIs, not POSIX header. Are posix APIs even needed for this sample?

lucasdietrich replied:
I didn't made the change yet, do you suggest to remove `#include <poll.h>` and `#include <unistd.h>`, then use `zsock_*` variants for poll related stuff ? It that case, we could almost disable `CONFIG_POSIX_API` (last POSIX function used is `clock_settime` from  `#include <time.h>`).

If we decide to pursue this approach, my suggestion would be to enable `CONFIG_POSIX_CLOCK` to still have the convenient function `clock_settime` through `#include <zephyr/posix/time.h>` rather than `#include <time.h>`.

What are your thoughts on this?

rlubos replied:
I'd just prefer to be consistent with the API usage, if started to use `zsock_*` APIs, just stick to it.

As for clock, enabling `CONFIG_POSIX_CLOCK` only seems fine.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+		clock_settime(CLOCK_REALTIME, &tspec);
+
+		LOG_DBG("Acquired time from NTP server: %u", (uint32_t)tspec.tv_sec);
+
```
Unneeded empty line.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+	};
+
+	ret = zsock_getaddrinfo(CONFIG_AWS_ENDPOINT, AWS_BROKER_PORT, &hints, &ai);
+
```
No need for an empty line here.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+	if (ret == 0) {
+		memcpy(broker, ai->ai_addr, MIN(ai->ai_addrlen, sizeof(struct sockaddr_storage)));
+
+		char addr_str[INET_ADDRSTRLEN];
```
It's more of a convention, but we aim to define variables at the top of the block.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+#define TLS_TAG_AWS_CA_CERTIFICATE 1
+
+static sec_tag_t sec_tls_tags[] = {
+	TLS_TAG_DEVICE_CERTIFICATE,
```
Ah, I've started to wonder how does it work if you only provide device certificate tag and no root CA, but now I see they're all regsitered under the same tag number.

I think it'd be a bit more readable, if root CA received its own tag, as it's not related to the device credentials. Device private tag and cert should share the same tag though, as they come in couple.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+static int subscribe_topic(void)
+{
+	int ret;
+
```
Those empty lines between variables defined are not needed.

lucasdietrich replied:
Fixed, do you identify these style inconsistencies on your own or do you have a script for it?

rlubos replied:
I'm afraid this is not verified by the scripts we have, so yes, manual spot.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+	}
+
+	fds.fd = client_ctx.transport.tcp.sock;
+	fds.events = POLLIN | POLLHUP | POLLERR;
```
POLLHUP and POLLERR are `revents` only - no need to register for them.

lucasdietrich replied:
Didn't know, fixed.

rlubos commented on samples/net/cloud/aws_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,494 @@
+					LOG_ERR("Failed to read MQTT input: %d", rc);
+					break;
+				}
+			}
```
I'd also check for POLLHUP and POLLERR here, and break if set.

rlubos reviewed (commented):
Thanks @lucasdietrich for the effort in contributing the sample. Looks pretty neat.

lucasdietrich commented on samples/net/cloud/aws_iot_mqtt/src/dhcp.c:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	for (i = 0; i < NET_IF_MAX_IPV4_ADDR; i++) {
+
```
See my comment on dhcp.c

rlubos reviewed (approved):
LGTM

lucasdietrich committed:
samples: net: mqtt: Add a sample to connect to AWS IoT core

This sample application demonstrates the implementation of an
MQTT client that can publish messages to AWS IoT Core using
the MQTT protocol.

Fixes #22866

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>
